File: Day 11/solution.ts
function minMutation(start: string, end: string, bank: string[]): number {
    const queue = new ArrayDeque<string>();
    const set = new Set<string>();
    const char = ['A', 'T', 'G', 'C'];
    var ans = 0;
    queue.offer(start);
    set.add(start);
    while(!queue.isEmpty())
        {
            var nodes = queue.size();
            for(let i = 0; i < nodes; i++)
                {
                    let currentNode = queue.poll();
                    if(currentNode === end)
                        {
                            return ans;
                        }
                    for(let c of char)
                        {
                            for(let j = 0; j < currentNode.length; j++)
                                {
                                    let mutation = currentNode.substring(0, j) + c + currentNode.substring(j + 1);
                                    if(bank.includes(mutation) && !set.has(mutation))
                                        {
                                            queue.offer(mutation);
                                            set.add(mutation);
                                        }
                                }
                        }
                }
            ans++;
        }
   
    return -1;
};
class ArrayDeque<T>
    {
        private buffer;
        public constructor()
        {
            this.buffer = new Array<T>();
        }
        public offer(item: T)
        {
            this.buffer.push(item);
        }
        public peek(): T
        {
            return this.buffer[0];
        }
        public poll(): T
        {
            const tmp: T = this.buffer[0];
            this.buffer.shift();
            return tmp;
        }
        public size(): number
        {
            return this.buffer.length;
        }
        public isEmpty(): boolean
        {
            return (this.buffer.length == 0);
        }
    }

File: Day 11/solution.java
class Solution {
    public int minMutation(String start, String end, String[] bank) {
        Queue<String> q = new LinkedList<>();
        HashSet<String> visited = new HashSet<>();
        char[] c = new char[]{'A', 'T', 'G', 'C'};
        q.offer(start);
        visited.add(start);
        int ans = 0;
        while(!q.isEmpty())
        {
            int nodes = q.size();
            for(int i = 0; i < nodes; i++)
            {
                String current = q.poll();
                if(current.equals(end))
                {
                    return ans;
                }
                StringBuilder mutation = new StringBuilder();
                for(char j : c)
                {
                    for(int k = 0; k < current.length(); k++)
                    {
                        mutation.append(current.substring(0, k));
                        mutation.append(j);
                        mutation.append(current.substring(k + 1));
                        if(!visited.contains(mutation.toString()) &&
                           Arrays.stream(bank).anyMatch(mutation.toString()::equals))
                        {
                            q.offer(mutation.toString());